### PR TITLE
Mesh geometry library and the mevegs user code 

### PR DESCRIPTION
**Pull request by mxxo**
**Summary**
This pull request open sources the mesh transport geometry library developed by Mevex Corporation. 
It is used at Mevex to model radiation transport through complex CAD geometries. 

We contribute it to the EGSnrc project to advance the state of the art for radiation modelling.     

**Update** 
This PR has been split into a number of smaller PRs to ease integration: 
1. Core mesh geometry (#637) 
2. Gmsh msh to `egs_mesh` translation (**TODO**)
3. Post-processing steps (**TODO**)

**TODO**
There are a number of blocking issues to fix before I'd consider this work ready to merge.
- [x] Address user code's Gmsh API dependency - [link](http://gmsh.info/doc/texinfo/gmsh.html#Gmsh-API)
- [ ] Comprehensive documentation with examples
- [x] Remove dependence on custom `egs_vector` operator `>` in favour of the standard `-` operator
- [x] Refactor heavy use of raw arrays in the `egs_mesh` constructor
- [x] Bring `egs_mesh` header and impl files in line with the other geometry libraries (declarations, etc) 
- [x] Polish the example user code `mevegs.cpp` 
- [ ] Format code using `astyle`, `clang-format` or similar 
- [ ] Static analysis (`UBSan`, `MemSan`, `CppCheck`, etc.) 
- [ ] Unit testing and benchmarks 
- [ ] Results comparison to non-meshed geometries 

**Copyright**
We contribute the code as an external collaborator under the terms of the AGPL3 license. 
Mevex retains code copyright.   

**Acknowledgements** 
This library is the product of many people's hard work at Mevex. 
Thank you to the entire team.

This library was built by:
- Dave Macrillo 
- Matt Ronan
- Nigel Vezeau 
- Lou Thompson
- Max Orok 

with contributions by Pascal Michaud. 

Special thanks to Dave Brown and Emily Craven for their support in publicly releasing this work.

### CODE REVIEW

**Comment by ftessier** [comment]
:medal_sports: to @mxxo and Mevex for their contribution to the EGSnrc free software project. We look forward to merging this pull request to implement CAD mesh transport, before the next release in 2020! :trophy:

**Comment by mxxo** [comment]
Thank you very much for your support! I am excited to get this pull request merged. 

One issue I've just added to the list is the `mevegs` user code's dependency on the Gmsh API ([version 4](http://gmsh.info/doc/texinfo/gmsh.html#Gmsh-API)). As a single user, it was easy to throw the Gmsh shared library into the linker path and upgrade whenever we needed to but having this code part of a standard release merits another look. 

Gmsh support the [three major platforms](http://gmsh.info/doc/texinfo/gmsh.html#Gmsh-API), but is it acceptable to have a standard user code fail to build if the library isn't installed? Do you have any advice about this? 

For CI, there's also some [provisioning issues](https://github.com/rust-lang/crates-build-env/issues/19) for building using standard releases since the Gmsh API is relatively recent.

**Comment by ftessier** [comment]
I propose to add Gmsh to the (optional) requirements to build EGSnrc. For GNU/Linux, I want to create .rpm and .deb packages that resolve dependencies automatically, including the Gmsh package (which I see is available in recent distros). I don't suggest we include Gmsh code in our repository (although that is possible given the GPL2+ licence), but then we need to keep it up to date, etc.

**Comment by mxxo** [comment]
That's a great idea, I also think that's a good way forward.

**Comment by ftessier** [comment]
@mxxo can you rebase this branch `feature-egs_mesh` on develop: `git rebase develop`. Post release I like to bring all active branches forward to tidy up and ensure that we don't accumulate too many conflicts over time!

**Comment by mxxo** [comment]
You got it Frédéric, thanks for the heads up!

File: HEN_HOUSE/scripts/configure_c++ [hunk 1]
```diff
@@ -165,7 +165,7 @@ if test $create_config=yes; then
 
 case $CXX in
 
-    *g++*) opt="-O2 -mtune=native";;# mingw32-g++ and g++4 also taken into account
```
Comment (rtownson) [inline HEN_HOUSE/scripts/configure_c++]: You'll also need to make this change in egs_configure/egs_tools.cpp to affect the GUI installer.

Comment (mxxo) [reply HEN_HOUSE/scripts/configure_c++]: Thank you Reid :)

Comment (mxxo) [reply HEN_HOUSE/scripts/configure_c++]: Tried to address using 7f67c92, wasn't sure about including corresponding flags for other compilers. Apparently recent versions `cl` don't have a `std=c++11` flag at all and the features are enabled by default: https://docs.microsoft.com/en-us/cpp/build/reference/std-specify-language-standard-version?view=vs-2019

**Comment by mxxo** [comment]
I spent the last couple of weeks refactoring the frontend and adding a python post-processing step. 
I don't think there's any major code changes left so am marking this ready for review 😄

**Comment by mxxo** [comment]
@ftessier I have been thinking about splitting this big PR into a few standalone smaller ones to ease review and ensure they fit with the rest of the code.

I think the three main pieces are
1. the mesh geometry (inside `egs++/geometry/egs_mesh`)
2. the `Gmsh` -> EGS mesh translation pipeline (`user_codes/mevegs` and `user_codes/mevegs/include`)
3. the post-processing steps (`user_codes/mevegs/postpro`)

**Comment by ftessier** [comment]
@mxxo sounds good, makes sense. Feel free to go ahead with creating the 3 smaller pull requests, referencing this one (which we will then close). I can also split it myself according to what you said, if you prefer!

**Comment by mxxo** [comment]
I'll go ahead and split them, thank you :)
The two outstanding things I'd like to change are: 
1. Take the mesh file as a `egsinp` input instead of a commandline arg
2. Refactor the tetrahedron geometry code as discussed

**Comment by rtownson** [comment]
I just wanted to note here that ICRP 145 includes mesh phantoms. That would be something to try to support in the future.

**Comment by mxxo** [comment]
> 
> 
> I just wanted to note here that ICRP 145 includes mesh phantoms. That would be something to try to support in the future.

Hey @rtownson , thank you for the heads up about that report! I just wrote a small parser using the new `EGS_Mesh` types from #637 to read ICRP node and element data files: https://gist.github.com/mxxo/df3c7d14a7aa941f2b2d6aa2f91ad2dc (no material data yet). 

It seems like it's working, though I've only done a quick smoke test using the `MCRP_AM` model files: 
```text
$ ./mrcp-parser MRCP_AM
MRCP_AM
read 1222593 nodes
node 1222592:
  x: -26.7065
  y: 10.1514
  z: -0.891105
read 8233413 tetrahedrons
element 8233412:
  organ_id: 802
  a: 18318
  b: 19476
  c: 18360
  d: 18319
```

After #637 lands maybe we can add support for the ICRP 145 files (maybe I should open a new issue for that). I think the tricky part here will be associating all the correct materials with the elements, whether we're going to have to make a special pegs4dat file, etc.

**Comment by ftessier** [comment]
Pushing this to Release 2022. We will want to merge this into `develop` as soon as possible after Release 2021!

**Comment by ftessier** [comment]
@mxxo I am merging in the mesh geometry for Release 2022. Can you confirm that this is the up to date branch for egs_mesh?

**Comment by ftessier** [comment]
Rebased on develop for Release 2022. Fixed compilation flag conflicts (this branch added the `-std=c++11` flag, but we have in the meantime added `-std=c++14`).

**Comment by mxxo** [comment]
Hi @ftessier, sorry for the delay, this is an outdated branch. The current `egs_mesh` branch is called `octree` (mxxo/octree), I will prepare a new PR.

**Comment by ftessier** [comment]
> Hi @ftessier, sorry for the delay, this is an outdated branch. The current `egs_mesh` branch is called `octree` (mxxo/octree), I will prepare a new PR.

Ah yes, you told me that many times, I had forgotten the name of the current branch. Thanks for clearing that up!

**Pull request closed without merging**